programs/marinade-finance/src/instructions/user/withdraw_stake_account.rs

```diff
@@ -0,0 +1,302 @@
+            StakeSystem::STAKE_WITHDRAW_SEED
+        ],
+        bump = state.stake_system.stake_withdraw_bump_seed
+    )]
+    pub stake_withdraw_authority: UncheckedAccount<'info>,
+    #[account(mut)]
+    pub stake_account: Box<Account<'info, StakeAccount>>,
+
+    #[account(
+        seeds = [
+            &state.key().to_bytes(),
+            StakeSystem::STAKE_DEPOSIT_SEED
+        ],
+        bump = state.stake_system.stake_deposit_bump_seed
+    )]
+    pub split_stake_account: Account<'info, StakeAccount>,
+
+    pub clock: Sysvar<'info, Clock>,
+    pub system_program: Program<'info, System>,
+    pub token_program: Program<'info, Token>,
+    pub stake_program: Program<'info, Stake>,
+}
+
+impl<'info> WithdrawStakeAccount<'info> {
+    pub fn process(
+        &mut self,
+        stake_index: u32,
+        validator_index: u32,
+        msol_amount: u64,
+    ) -> Result<()> {
+        require!(!self.state.paused, MarinadeError::ProgramIsPaused);
+        require!(!self.state.withdraw_stake_account_enabled, MarinadeError::WithdrawStakeAccountIsNotEnabled);
+
+        // record  for event
+        let user_msol_balance = self.burn_msol_from.amount;
+        // save msol price source
+        let total_virtual_staked_lamports = self.state.total_virtual_staked_lamports();
+        let msol_supply = self.state.msol_supply;
+
+        check_token_source_account(
+            &self.burn_msol_from,
+            self.burn_msol_authority.key,
+            msol_amount,
+        )?;
+
+        let mut stake = self.state.stake_system.get_checked(
+            &self.stake_list.data.as_ref().borrow(),
+            stake_index,
+            self.stake_account.to_account_info().key,
+        )?;
+        let last_update_stake_delegation = stake.last_update_delegated_lamports;
+
+        // require the stake is not in emergency_unstake
+        require_neq!(
+            stake.is_emergency_unstaking,
+            0,
+            MarinadeError::StakeAccountIsEmergencyUnstaking
+        );
+
+        // require stake is active (deactivation_epoch == u64::MAX)
+        let delegation = self.stake_account.delegation().ok_or_else(|| {
+            error!(MarinadeError::RequiredDelegatedStake).with_account_name("stake_account")
+        })?;
+        require_eq!(
+            delegation.deactivation_epoch,
+            std::u64::MAX,
+            MarinadeError::RequiredActiveStake
+        );
+
+        let mut validator = self
+            .state
+            .validator_system
+            .get(&self.validator_list.data.as_ref().borrow(), validator_index)?;
+
+        // check currently_staked in this account & validator vote-key
+        check_stake_amount_and_validator(
+            &self.stake_account,
+            stake.last_update_delegated_lamports,
+            &validator.validator_account,
+        )?;
+
+        // compute how many lamport the burned mSOL represents
+        require_gt!(msol_amount, self.state.min_withdraw);
```

Comment:
min withdraw is expressed in SOL not in mSOL. Do not confuse our reader

Reply:
ok, comparing with SOL
Q: can we create a slplit account with <= 1 SOL? or we weill have problems in future solana versions?